File: sdks/go/cmd/elydora/plugins/augment.go
package plugins

import (
	"fmt"
	"path/filepath"
)

// AugmentPlugin manages Auggie native global user hooks.
type AugmentPlugin struct {
	rename renameFunc
}

// ManagesGuardRuntime reports that Auggie commits its provider guard together
// with both wrappers, the audit runtime, and user settings.
func (p *AugmentPlugin) ManagesGuardRuntime() bool {
	return true
}

// PreflightInstall validates settings and runtime identity before any write.
func (p *AugmentPlugin) PreflightInstall(config InstallConfig) error {
	document, err := readAugmentDocument()
	if err != nil {
		return err
	}
	_, _, err = preflightAugmentInstallation(config, document)
	return err
}

func (p *AugmentPlugin) Install(config InstallConfig) error {
	document, err := readAugmentDocument()
	if err != nil {
		return err
	}
	paths, nodePath, err := preflightAugmentInstallation(config, document)
	if err != nil {
		return err
	}
	hooks, _ := removeManagedAugmentHooks(document.hooks, "", paths.runtimeRoot)
	hooks["PreToolUse"] = append(
		hooks["PreToolUse"],
		buildAugmentGroup(buildAugmentHandler(paths.guardWrapperPath)),
	)
	hooks["PostToolUse"] = append(
		hooks["PostToolUse"],
		buildAugmentGroup(buildAugmentHandler(paths.auditWrapperPath)),
	)
	rendered, err := renderAugmentDocument(document, hooks)
	if err != nil {
		return fmt.Errorf("render Auggie user settings: %w", err)
	}
	changes, err := prepareAugmentInstallationChanges(
		config,
		paths,
		nodePath,
		rendered,
	)
	if err != nil {
		return err
	}
	if err := writeAugmentChanges(
		changes,
		"Install Augment Code CLI hooks",
		p.rename,
		paths.runtimeRoot,
		paths.agentDirectory,
		filepath.Dir(document.configPath),
	); err != nil {
		return err
	}
	fmt.Println("Auggie: user-level PreToolUse and PostToolUse hooks installed.")
	return nil
}

func (p *AugmentPlugin) Uninstall(agentID string) error {
	document, err := readAugmentDocument()
	if err != nil {
		return err
	}
	if !document.exists {
		return nil
	}
	runtimeRoot, err := AgentRuntimeRoot()
	if err != nil {
		return err
	}
	hooks, changed := removeManagedAugmentHooks(
		document.hooks,
		agentID,
		runtimeRoot,
	)
	if !changed {
		return nil
	}
	rendered, err := renderAugmentDocument(document, hooks)
	if err != nil {
		return fmt.Errorf("render Auggie user settings: %w", err)
	}
	change, err := prepareRenderedAugmentChange(rendered)
	if err != nil {
		return err
	}
	return writeAugmentChanges(
		[]*fileChange{change},
		"Uninstall Augment Code CLI hooks",
		p.rename,
		"",
		"",
		filepath.Dir(document.configPath),
	)
}

func (p *AugmentPlugin) Status() (PluginStatus, error) {
	configPath, pathErr := augmentConfigPath()
	entry := SupportedAgents[augmentAgentKey]
	status := PluginStatus{
		AgentName: augmentAgentKey, DisplayName: entry.Name, ConfigPath: configPath,
	}
	if pathErr != nil {
		return status, pathErr
	}
	document, err := readAugmentDocument()
	if err != nil {
		return status, err
	}
	runtimeRoot, err := AgentRuntimeRoot()
	if err != nil {
		return status, err
	}
	contracts := augmentRuntimeContracts(document.hooks, runtimeRoot)
	status.HookConfigured = len(contracts) > 0
	if !status.HookConfigured {
		return status, nil
	}
	status.HookScriptExists, err = augmentRuntimeFilesExist(
		contracts,
		runtimeRoot,
	)
	if err != nil {
		return status, err
	}
	status.Installed = status.HookScriptExists
	return status, nil
}

File: sdks/node/src/plugins/augment.ts
import type { AgentPlugin, InstallConfig, PluginStatus } from './base.js';
import { SUPPORTED_AGENTS } from './registry.js';
import {
  AGENT_KEY,
  buildHandler,
  removeManagedHooks,
  renderAugmentDocument,
  runtimeContracts,
  wrapperPaths,
} from './augment-contract.js';
import {
  commitAugmentInstallation,
  preflightAugmentInstallation,
  prepareAugmentInstallation,
} from './augment-installation.js';
import {
  augmentRuntimeFilesExist,
  readConfig,
  writeAugmentDocument,
} from './augment-io.js';

const entry = SUPPORTED_AGENTS.get(AGENT_KEY)!;

export const augmentPlugin: AgentPlugin = {
  managesRuntime: true,

  async preflightInstall(config: InstallConfig): Promise<void> {
    const document = await readConfig();
    await preflightAugmentInstallation(config, document.configPath);
  },

  async install(config: InstallConfig): Promise<void> {
    const document = await readConfig();
    const paths = await preflightAugmentInstallation(config, document.configPath);
    const wrappers = wrapperPaths(paths.agentDirectory);

    const cleaned = removeManagedHooks(document.hooks).hooks;
    const hooks = {
      ...cleaned,
      PreToolUse: [
        ...(cleaned.PreToolUse ?? []),
        { matcher: '.*', hooks: [buildHandler(wrappers.guardPath)] },
      ],
      PostToolUse: [
        ...(cleaned.PostToolUse ?? []),
        { matcher: '.*', hooks: [buildHandler(wrappers.auditPath)] },
      ],
    };
    const rendered = renderAugmentDocument(document, hooks);
    const prepared = await prepareAugmentInstallation(config, rendered);
    await commitAugmentInstallation(prepared);
    console.log('  Auggie: user-level PreToolUse and PostToolUse hooks installed.');
  },

  async uninstall(agentId?: string): Promise<void> {
    const document = await readConfig();
    if (!document.exists) return;
    const result = removeManagedHooks(document.hooks, agentId);
    if (!result.changed) return;
    await writeAugmentDocument(renderAugmentDocument(document, result.hooks));
  },

  async status(): Promise<PluginStatus> {
    const document = await readConfig();
    const contracts = runtimeContracts(document.hooks);
    const hookConfigured = contracts.length > 0;
    const hookScriptExists = hookConfigured
      ? await augmentRuntimeFilesExist(contracts)
      : false;
    return {
      installed: hookConfigured && hookScriptExists,
      agentName: AGENT_KEY,
      displayName: entry.name,
      hookConfigured,
      hookScriptExists,
      configPath: document.configPath,
    };
  },
};
